### PR TITLE
Kessel inventory client migration

### PR DESCRIPTION
**Pull request by bonscji1**
Assisted-by: Claude Sonnet 5 (via Claude Code)

## Summary by CodeRabbit

* **New Features**
  * Added bundle creation and editing in the administration console.
  * Added event-type linking and unlinking for behavior groups.
  * Added v2 subscription configuration APIs with filtering, severity selection, and validation.
  * Added Migration Advisor notification email templates.
* **Enhancements**
  * Event searches now support precise date and date-time ranges.
  * Large event exports are processed more efficiently through pagination.
  * Recipient lookups now use Kessel Inventory streaming with automatic retries and recovery.
* **Bug Fixes**
  * Fixed OpenAPI schema handling for array request bodies.
  * Improved email sender selection and event export reliability.

### CODE REVIEW

File: recipients-resolver/src/main/java/com/redhat/cloud/notifications/recipients/resolver/kessel/KesselService.java [hunk 1]
```diff
@@ -29,100 +28,47 @@ public class KesselService {
...
+    KesselInventoryClient kesselInventoryClient;
+
+    // worst-case retry budget (4 x kessel.timeout-ms = 120s) exceeds callers' default 30s REST timeout.
+    // Confirmed: none of engine/connector-email/connector-drawer override quarkus.rest-client.recipients-resolver's
+    // read-timeout, so all three genuinely get Quarkus's 30s default. Also, this budget isn't competing with an
+    // otherwise-empty request: RecipientsResolver.findRecipients() calls this lookup first, then still has to run
+    // RBAC/MBOP calls (FetchUsersFromExternalServices) per RecipientSettings in the same request/timeout window --
+    // so the safe target is meaningfully less than 30s, not just-under-30s. No real traffic exercises this path yet
+    // (use-kessel toggle is off everywhere), so there's no latency data to size against. Needs sizing guidance.
+    @Retry(maxRetries = 3, delay = 100, retryOn = KesselTransientException.class)
```
Comment (bonscji1) [inline recipients-resolver/src/main/java/com/redhat/cloud/notifications/recipients/resolver/kessel/KesselService.java:41]: This is problematic, i would appreciate any help.

File: recipients-resolver/pom.xml [hunk 2]
```diff
@@ -34,8 +34,14 @@
...
-            <version>${kessel.relationships-client.version}</version>
+            <artifactId>kessel-sdk</artifactId>
+            <version>${kessel-sdk.version}</version>
+        </dependency>
+
+        <!-- gRPC transport for the Kessel client (kessel-sdk does not bring one transitively) -->
+        <dependency>
+            <groupId>io.grpc</groupId>
+            <artifactId>grpc-netty-shaded</artifactId>
         </dependency>
```
[the remark this answers is not in the document]
Comment (bonscji1) [reply recipients-resolver/pom.xml:45]: ran dependency:tree filtered on grpc-netty-shaded, and it resolves cleanly to 1.81.0 via the inherited quarkus-bom
  (dependencyManagement in the root pom just imports quarkus-bom, which already version-manages this artifact)

**Comment by larrymou9** [comment]
✅ **Performance Tests passed**

| | |
|---|---|
| **PR** | #4728 |
| **Branch** | `RHCLOUD-44097-kessel-api-migration` |
| **SHA** | `40c2e8f75c1543703cd540e2d43818cac826e29a` |
| **Build** | [#32](https://jenkins-csb-perf-master.dno.corp.redhat.com/job/InsightsNotifications_PRTest_builder/32/) |
| **Image tested** | `quay.io/redhat-user-workloads/hcc-integrations-tenant/notifications/notifications-backend:on-pr-40c2e8f75c1543703cd540e2d43818cac826e29a` |

Tests run: InsightsNotificationswebhook_runner · InsightsNotificationsgw_runner · InsightsNotificationsemail_runner

File: recipients-resolver/src/main/java/com/redhat/cloud/notifications/recipients/resolver/kessel/KesselInventoryClient.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    public static final String KESSEL_GRPC_ERROR_TAG_ERROR_TYPE = "error_type";
+
+    private static final List<Status.Code> TRANSIENT_FAILURE_CODES = List.of(UNAVAILABLE, DEADLINE_EXCEEDED, RESOURCE_EXHAUSTED, ABORTED);
+    private static final long CHANNEL_SHUTDOWN_TIMEOUT_SECONDS = 30;
+
+    @Inject
+    MeterRegistry meterRegistry;
+
+    @Inject
+    OAuth2ClientCredentialsCache oauth2ClientCredentialsCache;
```
Comment (g-duval) [inline recipients-resolver/src/main/java/com/redhat/cloud/notifications/recipients/resolver/kessel/KesselInventoryClient.java:62]: Why do we need to keep those credentials in cache? can't we handle those in `postConstruct` method?

File: recipients-resolver/src/main/java/com/redhat/cloud/notifications/recipients/resolver/kessel/KesselService.java (hunk 1 above)
Comment (g-duval) [reply recipients-resolver/src/main/java/com/redhat/cloud/notifications/recipients/resolver/kessel/KesselService.java:41]: We will revisit it later, once https://github.com/RedHatInsights/notifications-backend/pull/4539 will be merged.

**Review by g-duval — COMMENTED** [review]
Looks good, could you please check if we can avoid the cacheing credentials?

File: recipients-resolver/src/main/java/com/redhat/cloud/notifications/recipients/resolver/kessel/KesselInventoryClient.java (hunk 3 above)
Comment (bonscji1) [reply recipients-resolver/src/main/java/com/redhat/cloud/notifications/recipients/resolver/kessel/KesselInventoryClient.java:62]: This handling is something copied/borrowed from backend. On further discussion with claude, i believe the rationale is that postConstruct only runs once during the bean lifetime, while the current cashe is invalidated and refreshed on: channel init, unheathy channel recovery and UNAUTHENTICATED grpc error which gives us free edge case error handling we would likely have to do manually with postConstruct.

**Comment by larrymou9** [comment]
❌ **Performance Tests failed**

| | |
|---|---|
| **PR** | #4728 |
| **Branch** | `RHCLOUD-44097-kessel-api-migration` |
| **SHA** | `1eb1550c2d9f3b6417d3106d9ccca58f8729e881` |
| **Build** | [#103](https://jenkins-csb-perf-master.dno.corp.redhat.com/job/InsightsNotifications_PRTest_builder/103/) |
| **Image tested** | `quay.io/redhat-user-workloads/hcc-integrations-tenant/notifications/notifications-backend:on-pr-1eb1550c2d9f3b6417d3106d9ccca58f8729e881` |

Tests run: InsightsNotificationswebhook_runner · InsightsNotificationsgw_runner · InsightsNotificationsemail_runner

<details>
<summary>📊 Performance Results</summary>

**InsightsNotificationsemail_runner**

| Metric | Result | Value | Min | Max |
|--------|--------|-------|-----|-----|
| `notifications-engine-service.restarts.sum` | ✅ PASS | 0.000 | 0.000 | 0.000 |
| `notifications-connector-email-service.restarts.sum` | ✅ PASS | 0.000 | 0.000 | 0.000 |
| `SUMMARY.total_rps` | ❌ FAIL | 14.502 | 13.966 | 14.498 |
| `SUMMARY.median_response_time` | ✅ PASS | 12.000 | 10.000 | 12.000 |
| `POST_notifications.avg_response_time` | ❌ FAIL | 24.981 | 25.972 | 30.650 |
| `SUMMARY.num_failures` | ✅ PASS | 0.000 | 0.000 | 0.000 |
| `SUMMARY.fail_ratio` | ✅ PASS | 0.000 | 0.000 | 0.000 |

**InsightsNotificationswebhook_runner**

| Metric | Result | Value | Min | Max |
|--------|--------|-------|-----|-----|
| `notifications-engine-service.restarts.sum` | ✅ PASS | 0.000 | 0.000 | 0.000 |
| `notifications-connector-webhook-service.restarts.sum` | ✅ PASS | 0.000 | 0.000 | 0.000 |
| `results.created_at.duration_stats.mean` | ❌ FAIL | 155.377 | 141.021 | 154.384 |
| `results.created_at.duration_stats.median` | ❌ FAIL | 171.718 | 152.376 | 170.057 |
| `results.created_at.rps_stats.mean` | ✅ PASS | 3.464 | 3.240 | 3.678 |
| `results.created_at.rps_stats.median` | ✅ PASS | 2.900 | 2.900 | 3.200 |

**InsightsNotificationsgw_runner**

| Metric | Result | Value | Min | Max |
|--------|--------|-------|-----|-----|
| `notifications-engine-service.restarts.sum` | ✅ PASS | 0.000 | 0.000 | 0.000 |
| `notifications-gw-service.restarts.sum` | ✅ PASS | 0.000 | 0.000 | 0.000 |
| `SUMMARY.total_rps` | ✅ PASS | 14.329 | 14.105 | 14.599 |
| `SUMMARY.median_response_time` | ✅ PASS | 10.000 | 10.000 | 12.000 |
| `POST_notifications.avg_response_time` | ✅ PASS | 26.966 | 24.871 | 29.947 |
| `SUMMARY.num_failures` | ✅ PASS | 0.000 | 0.000 | 0.000 |
| `SUMMARY.fail_ratio` | ✅ PASS | 0.000 | 0.000 | 0.000 |

</details>

File: admin-console/src/main/webapp/src/app/App.tsx [hunk 4]
```diff
@@ -33,6 +37,46 @@ export const App: React.FunctionComponent<unknown> = () => {
...
+        }).then((response) => {
+            if (response.error) {
+                setBundleCreateError('Failed to create bundle. Please check the values and try again.');
+            } else {
+                setShowBundleModal(false);
+                bundles.query();
+                const createdBundle = response.payload?.type === 'Bundle' ? response.payload.value : undefined;
+                if (createdBundle?.id) {
+                    navigate(linkTo.bundle(createdBundle.id));
+                }
```
[the remark this answers is not in the document]
Comment (bonscji1) [reply admin-console/src/main/webapp/src/app/App.tsx:72]: not part of this PR concern, i will mark it separatedly

File: admin-console/src/main/webapp/src/components/SystemBehaviorGroups/BehaviorGroupEventTypesPanel.tsx [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    const [ checkedIds, setCheckedIds ] = React.useState<Set<string>>(new Set());
+    const [ saving, setSaving ] = React.useState(false);
+    const [ saveError, setSaveError ] = React.useState<string | null>(null);
+    const [ saveSuccess, setSaveSuccess ] = React.useState(false);
+
+    React.useEffect(() => {
+        if (!selectedAppId) {
+            setEventTypes([]);
+            setCheckedIds(new Set());
+            return;
```
[the remark this answers is not in the document]
Comment (bonscji1) [reply admin-console/src/main/webapp/src/components/SystemBehaviorGroups/BehaviorGroupEventTypesPanel.tsx:46]: not part of the PR, will mark separatedly

File: admin-console/src/main/webapp/src/components/SystemBehaviorGroups/BehaviorGroupEventTypesPanel.tsx [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+        };
+    }, [ selectedAppId, query, props.behaviorGroup ]);
+
+    const hasChanges = React.useMemo(() => {
+        return eventTypes.some(et => {
+            const wasLinked = isEventTypeLinked(props.behaviorGroup, et.id);
+            const isChecked = checkedIds.has(et.id);
+            return wasLinked !== isChecked;
+        });
+    }, [ eventTypes, checkedIds, props.behaviorGroup ]);
```
[the remark this answers is not in the document]
Comment (bonscji1) [reply admin-console/src/main/webapp/src/components/SystemBehaviorGroups/BehaviorGroupEventTypesPanel.tsx:117]: admin changes are separate from PR, it comes from rebase

**Comment by larrymou9** [comment]
❌ **Performance Tests failed**

| | |
|---|---|
| **PR** | #4728 |
| **Branch** | `RHCLOUD-44097-kessel-api-migration` |
| **SHA** | `b056682fdc72b63f1d6e0565244214623d8f0c1c` |
| **Build** | [#104](https://jenkins-csb-perf-master.dno.corp.redhat.com/job/InsightsNotifications_PRTest_builder/104/) |
| **Image tested** | `quay.io/redhat-user-workloads/hcc-integrations-tenant/notifications/notifications-backend:on-pr-b056682fdc72b63f1d6e0565244214623d8f0c1c` |

Tests run: InsightsNotificationswebhook_runner · InsightsNotificationsgw_runner · InsightsNotificationsemail_runner

<details>
<summary>📊 Performance Results</summary>

**InsightsNotificationsemail_runner**

| Metric | Result | Value | Min | Max |
|--------|--------|-------|-----|-----|
| `notifications-engine-service.restarts.sum` | ✅ PASS | 0.000 | 0.000 | 0.000 |
| `notifications-connector-email-service.restarts.sum` | ✅ PASS | 0.000 | 0.000 | 0.000 |
| `SUMMARY.total_rps` | ❌ FAIL | 14.584 | 13.966 | 14.498 |
| `SUMMARY.median_response_time` | ✅ PASS | 10.000 | 10.000 | 12.000 |
| `POST_notifications.avg_response_time` | ❌ FAIL | 25.135 | 25.972 | 30.650 |
| `SUMMARY.num_failures` | ✅ PASS | 0.000 | 0.000 | 0.000 |
| `SUMMARY.fail_ratio` | ✅ PASS | 0.000 | 0.000 | 0.000 |

**InsightsNotificationswebhook_runner**

| Metric | Result | Value | Min | Max |
|--------|--------|-------|-----|-----|
| `notifications-engine-service.restarts.sum` | ✅ PASS | 0.000 | 0.000 | 0.000 |
| `notifications-connector-webhook-service.restarts.sum` | ✅ PASS | 0.000 | 0.000 | 0.000 |
| `results.created_at.duration_stats.mean` | ✅ PASS | 152.838 | 141.021 | 154.384 |
| `results.created_at.duration_stats.median` | ✅ PASS | 169.707 | 152.376 | 170.057 |
| `results.created_at.rps_stats.mean` | ✅ PASS | 3.642 | 3.240 | 3.678 |
| `results.created_at.rps_stats.median` | ✅ PASS | 3.200 | 2.900 | 3.200 |

**InsightsNotificationsgw_runner**

| Metric | Result | Value | Min | Max |
|--------|--------|-------|-----|-----|
| `notifications-engine-service.restarts.sum` | ✅ PASS | 0.000 | 0.000 | 0.000 |
| `notifications-gw-service.restarts.sum` | ✅ PASS | 0.000 | 0.000 | 0.000 |
| `SUMMARY.total_rps` | ✅ PASS | 14.499 | 14.105 | 14.599 |
| `SUMMARY.median_response_time` | ✅ PASS | 12.000 | 10.000 | 12.000 |
| `POST_notifications.avg_response_time` | ✅ PASS | 26.102 | 24.871 | 29.947 |
| `SUMMARY.num_failures` | ✅ PASS | 0.000 | 0.000 | 0.000 |
| `SUMMARY.fail_ratio` | ✅ PASS | 0.000 | 0.000 | 0.000 |

</details>

**Review by g-duval — APPROVED** [review]
LGTM
